linode/manager
# refactor: [M3-7380] - Use `volumes/types` endpoint for pricing data

## Description 📝
API is now returning pricing data through the `/volumes/types` endpoint. 

In this PR, prices can be dynamically displayed for base region pricing and DC-specific region pricing, using the prices returned in the new endpoint, for Volumes.

## Changes  🔄
- Created the necessary api-v4 endpoint, query, and interface to use the volumes/types endpoint.
- Modified `getDCSpecificPriceByType` to optionally accept a size param, if size is a factor of price and updated util unit test accordingly.
- Created the necessary factory and endpoint in the MSW to support a mocked response to the API endpoint.
- Added UX for error handling in the unlikely event that pricing cannot be calculated: disable create button and display tooltip error. 
- Added loading state for display prices.
- Updated `create-volume-smoke.spec.ts` to test create volume error flows.

Note: I did *not* update the resize or clone Cypress tests to test those error flows. These existing specs run slowly since they have to create a linode and I'm not sure it's worth it for an edge-case like this, so held off doing that and will talk to Joe when he returns.

## Preview 📷
The **only** visual changes should be loading state for pricing and the disabled Create Volume, Resize Volume, and Clone Volume buttons *if* the price is invalid (e.g. undefined).
| Flow | Screenshot |
| --- | --- |
| Create Volume Landing Error State | ![Screenshot 2024-04-12 at 12 17 57 PM](https://github.com/linode/manager/assets/114685994/08639d10-07be-473d-b5b7-11c3eb5e4892) |
| Create Volume from Linode Error State | ![Screenshot 2024-04-12 at 12 18 18 PM](https://github.com/linode/manager/assets/114685994/d4652403-4088-41f8-980b-a6ae960d04bf) |
| Resize Volume Error State | ![Screenshot 2024-04-12 at 12 18 43 PM](https://github.com/linode/manager/assets/114685994/aed5d967-6048-48a5-850e-03c6d7f604bf) |
| Clone Volume Error State | ![Screenshot 2024-04-12 at 10 41 33 AM](https://github.com/linode/manager/assets/114685994/0a9469f6-4f8f-463a-8030-752cc54057ad) |

## How to test 🧪

### Prerequisites
(How to setup test environment)
- Check out this PR and `yarn up`.
- All new `types` endpoints are all in prod, so use the prod environment.
- Have an existing linode or create one. 

### Verification steps
(How to verify changes)
- Go to http://localhost:3000/volumes/create.
- Fill out the form.
- Verify price calculation is correct by testing different options for the Region and comparing with prod. 
- Confirm:
   - There are no regressions in decimal point value or calculations displayed in the UI.
   - $0 is a valid region price if returned from the API, the user sees $0 pricing, and a volume can be created. (Update the mock data to test this.)
   - A user can still create a Volume.
   - If price cannot be calculated (e.g. it is undefined), the Summary displays the unknown price ($--.--) string, the Create button is disabled, and a tooltip displays on hover. (Blocking network requests is one method.)
   - When a region is not yet selected, the Create button is still enabled. (Pricing can't be calculated yet; we don't know if the price is invalid.)
   - Loading state functions as expected. 
- Additionally, confirm the above when creating or resizing a volume _from a linode_:
   - via Linode Details > Storage > Create Volume
   -  via Linode Details > Storage > Clone (the volume you created in the last step)
   - via Linode Details > Storage > Resize (the volume you created)


## As an Author I have considered 🤔

*Check all that apply*
